lib/async_hooks.js

```diff
@@ -300,10 +313,14 @@ class AsyncLocalStorage {
   }
 
   // Propagate the context from a parent resource to a child one
-  _propagate(resource, triggerResource) {
+  _propagate(resource, triggerResource, type) {
     const store = triggerResource[this.kResourceStore];
     if (this.enabled) {
-      resource[this.kResourceStore] = store;
+      if (this._onPropagateCb === null || this._onPropagateCb(type, store)) {
+        resource[this.kResourceStore] = store;
+      } else {
+        resource[this.kResourceStore] = undefined;
```

Comment:
Here I'm not completely sure if we should set `undefined` or skip setting the property and keep the state as it is.

I'm in general unsure why we propagate `undefined` at all but that's not changed in this PR.

Reply:
`undefined` is considered a valid value to propagate as a store may want to be expressed as _active_ but not currently containing a value. This is why we use `this.enabled` checks rather than `typeof` checks. The _correct_ thing to do would be to not propagate the field at all and do a `ObjectHasOwnProperty` check on the `triggerResource` to identify if the value should try to propagate.

Reply:
I removed setting `undefined` if propagation is disabled.

As a user of `AsyncLocalStorage` there is only `getStore()` which returns `undefined` if there is no store or if `undefined` is the propagated store. Anyhow this part of `AsyncLocalStorage` is not changed in this PR so lets tackle this outside of this PR.

Reply:
You're meant to check storage.enabled to identify if the store is active, not if getStore() returns undefined.

Reply:
`als.enabled()` at propagation time might be different to read time. But well, I guess noone will use `undefined` as store and enable/disable a store regularly....